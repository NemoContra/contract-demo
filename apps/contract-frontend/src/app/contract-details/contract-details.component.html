<ng-container *ngIf="isError$ | async as errorCode; else success">
  <div class="error">Search failed with HTTP-Error {{ errorCode }}</div>
</ng-container>
<ng-template #success>
  <ng-container *ngIf="loading$ | async; else content">
    <div class="loading">
      <mat-spinner></mat-spinner>
    </div>
  </ng-container>
  <ng-template #content>
    <ng-container *ngIf="data$ | async as data">
      <section class="title-container mat-elevation-z8">
        <h1>{{data.contractNumber.formatted}}</h1>
        <a mat-button routerLink="..">go to overview</a>
      </section>

      <section class="data-container">
        <p>{{data.description}}</p>
      </section>
    </ng-container>
  </ng-template>
</ng-template>
