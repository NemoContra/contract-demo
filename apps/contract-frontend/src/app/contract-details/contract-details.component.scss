.title-container {
  display: flex;
  align-items: center;

  padding: 5px 10px;
  margin-bottom: 20px;

  h1 {
    margin-bottom: 0;
  }

  a {
    margin-left: auto;
  }
}

.loading {
  display: flex;
  justify-content: center;
}
