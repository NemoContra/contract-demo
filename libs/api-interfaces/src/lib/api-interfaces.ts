export interface ContractFilterQuery {
  term?: string;
}

export interface Contract {
  id: string;
  person: {
    firstname: string;
    lastname: string;
  };
  type: ContractType;
}

export type ContractType = 'bAV' | 'Krankenversicherung' | 'KFZ-Versicherung';

export interface FilterResult {
  contracts: Contract[];
  totalElements: number;
}

export interface GroupcontractOverviewData {
  contractNumber: ContractNumber;
  oldContractNumbers: ContractNumber[];
  description: string;
}

export interface ContractNumber {
  id: string;
  formated: string;
}

