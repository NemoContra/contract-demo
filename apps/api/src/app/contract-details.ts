import { GroupcontractOverviewData } from '@contract-demo/api-interfaces';

export const contractDetailsData: GroupcontractOverviewData =
{
  vertragsnummer: {
    id: '1',
    formatiert: 'ID-1'
  },
  vorvertragsnummern: [
    {
      id: 'A1',
      formatiert: 'Old-ID-1'
    },
    {
      id: 'A2',
      formatiert: 'Old-ID-2'
    }
  ],
  vertragsbezeichnung: 'this is a nice description of the data.'
};
