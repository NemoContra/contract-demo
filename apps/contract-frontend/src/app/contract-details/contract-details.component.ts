import { Component, OnInit } from '@angular/core';
import { ContractDetailsFacade } from "./+state/contract-details.facade";
import { Observable } from "rxjs";
import { GroupcontractOverviewData } from "@contract-demo/api-interfaces";

@Component({
  selector: 'contract-demo-contract-details',
  templateUrl: './contract-details.component.html',
  styleUrls: ['./contract-details.component.scss']
})
export class ContractDetailsComponent implements OnInit {

  data$: Observable<GroupcontractOverviewData | undefined> = this.contractDetailsFacade.contractDetailsData$;
  loading$: Observable<boolean> = this.contractDetailsFacade.getContractDetailsLoading$;
  isError$: Observable<number | undefined> = this.contractDetailsFacade.getContractDetailsError$;
  constructor(private readonly contractDetailsFacade: ContractDetailsFacade) {}

  ngOnInit(): void {
    this.contractDetailsFacade.getContractDetailsData();
  }
}
