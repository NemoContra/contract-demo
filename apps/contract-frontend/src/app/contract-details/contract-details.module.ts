import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ContractDetailsComponent } from './contract-details.component';
import { StoreModule } from '@ngrx/store';
import { RouterModule } from "@angular/router";
import { EffectsModule } from "@ngrx/effects";
import { ContractDetailsEffects } from "./+state/contract-details.effects";
import { CONTRACT_DETAILS_FEATURE_KEY, contractDetailsReducer } from "./+state/contract-details.reducer";
import { MatProgressSpinnerModule } from "@angular/material/progress-spinner";
import { MatButtonModule } from "@angular/material/button";

@NgModule({
  imports: [
    CommonModule,
    MatProgressSpinnerModule,
    MatButtonModule,
    RouterModule.forChild([{ path: '', component: ContractDetailsComponent }]),
    StoreModule.forFeature(CONTRACT_DETAILS_FEATURE_KEY, contractDetailsReducer),
    EffectsModule.forFeature([ContractDetailsEffects]),
  ],
  declarations: [
    ContractDetailsComponent
  ],
})
export class ContractDetailsModule { }
