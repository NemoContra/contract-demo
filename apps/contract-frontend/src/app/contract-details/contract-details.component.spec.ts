import { ContractDetailsComponent } from './contract-details.component';
import { HarnessLoader } from "@angular/cdk/testing";
import { Subject } from "rxjs";
import { GroupcontractOverviewData } from "@contract-demo/api-interfaces";
import { createRoutingFactory, Spectator } from "@ngneat/spectator/jest";
import { MatProgressSpinnerModule } from "@angular/material/progress-spinner";
import { TestbedHarnessEnvironment } from "@angular/cdk/testing/testbed";
import { MatProgressSpinnerHarness } from "@angular/material/progress-spinner/testing";
import { ContractDetailsFacade } from "./+state/contract-details.facade";

describe('ContractDetailsComponent', () => {
  let spectator: Spectator<ContractDetailsComponent>;
  let loader: HarnessLoader;

  const contractDetailsFacadeMock = {
    getContractDetailsData: jest.fn(),
    contractDetailsData$: new Subject<GroupcontractOverviewData | undefined>(),
    getContractDetailsLoading$: new Subject<boolean>(),
    getContractDetailsError$: new Subject<number | undefined>(),
  };

  const createComponent = createRoutingFactory({
    component: ContractDetailsComponent,
    imports: [
      MatProgressSpinnerModule,
    ],
    providers: [
      {
        provide: ContractDetailsFacade,
        useValue: contractDetailsFacadeMock,
      },
    ],
    detectChanges: false,
  });

  it('should show correct data', () => {
    spectator = createComponent();
    spectator.detectChanges();
    contractDetailsFacadeMock.contractDetailsData$.next(mockContractDetailsData);
    contractDetailsFacadeMock.getContractDetailsLoading$.next(false);
    contractDetailsFacadeMock.getContractDetailsError$.next(undefined);
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();

    expect(spectator.query('.title-container h1')).toHaveText('ID-1');
    expect(spectator.query('.data-container p')).toHaveText('this is a nice description of the data.');
  });

  it('should show a loading spinner while content is loading', async () => {
    spectator = createComponent();
    loader = TestbedHarnessEnvironment.loader(spectator.fixture);
    spectator.detectChanges();
    contractDetailsFacadeMock.getContractDetailsLoading$.next(true);
    spectator.detectChanges();

    expect(await loader.getHarness(MatProgressSpinnerHarness)).not.toBeNull();

    contractDetailsFacadeMock.getContractDetailsLoading$.next(false);
    spectator.detectChanges();

    expect(spectator.query('mat-spinner')).toBeNull();
  });

  it('should show correct error message', () => {
    spectator = createComponent();
    spectator.detectChanges();
    contractDetailsFacadeMock.getContractDetailsError$.next(500);
    spectator.detectChanges();

    expect(spectator.query('.title-container')).toBeNull();
    expect(spectator.query('.data-container')).toBeNull();
    expect(spectator.query('.loading')).toBeNull();
    expect(spectator.query('.error')?.textContent).toEqual(
      'Search failed with HTTP-Error 500'
    );
  });
});

const mockContractDetailsData: GroupcontractOverviewData =
  {
    contractNumber: {
      id: '1',
      formatted: 'ID-1'
    },
    previousContractNumbers: [
      {
        id: 'A1',
        formatted: 'Old-ID-1'
      },
      {
        id: 'A2',
        formatted: 'Old-ID-2'
      }
    ],
    description: 'this is a nice description of the data.'
  };
