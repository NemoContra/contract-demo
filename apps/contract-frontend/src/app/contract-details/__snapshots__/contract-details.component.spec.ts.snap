// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractDetailsComponent should show correct data 1`] = `
<contract-demo-contract-details
  contractDetailsFacade={[Function Object]}
  data$={[Function Subject]}
  isError$={[Function Subject]}
  loading$={[Function Subject]}
>
  <section
    class="title-container mat-elevation-z8"
  >
    <h1>
      ID-1
    </h1>
    <a
      href="/"
      mat-button=""
      routerlink=".."
    >
      go to overview
    </a>
  </section><section
    class="data-container"
  >
    <p>
      this is a nice description of the data.
    </p>
  </section>
</contract-demo-contract-details>
`;
